spring-kafka/src/main/java/org/springframework/kafka/support/RequestReplyFuture.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.kafka.support;
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+
+import org.springframework.util.concurrent.ListenableFuture;
+import org.springframework.util.concurrent.SettableListenableFuture;
+
+/**
+ * A listenable future for requests/replies.
+ *
+ * @param <K> the key type.
+ * @param <V> the outbound data type.
+ * @param <R> the reply data type.
+ *
+ * @author Gary Russell
+ * @since 2.1.3
+ *
+ */
+public class RequestReplyFuture<K, V, R> extends SettableListenableFuture<ConsumerRecord<K, R>> {
+
+	private volatile ListenableFuture<SendResult<K, V>> sendFuture;
+
+	public RequestReplyFuture() {
+		super();
+	}
+
+	protected void setSendFuture(ListenableFuture<SendResult<K, V>> sendFuture) {
```

Comment:
I don't want the user to be able to set it.

Reply:
Yes, good point! Why do we need this class on the top level at all?
What doesn't work keeping it as inner one?

Thanks

Reply:
Where do you see a tangle? S101 doesn't see it. We can't use an inner class because then we can't have the ...Operations interface. I agree it should be moved though.

Reply:
Yeah... I think there was a tangle before, when we didn't have a new `requestreply`.

But that's good that it is fixed now !